### ISSUE TITLE
Suggestions:  

### ISSUE THREAD

**Issue by maxreiss123** [janneshopman/RKSymFoam#6]
Hi,

Thanks for that really good solver enhancement. I tried it using my own developed Non-Linear models. I found that the solver routine does not call that standard function `turbulence->divDevRhoReff(U)`. 

Because the BasicModel returns the following:

`- fvm::laplacian(this->alpha_*this->rho_*this->nuEff(), U)`

I would suggest changing:

https://github.com/janneshopman/RKSymFoam/blob/4254af93da39e10473cef3da7c59200e97b3cc1b/applications/solvers/RKSymFoam/store.H#L21

With:
```
        UEqnSourceB = - turbulence->divDevRhoReff(U);
```

Best regards

**Comment by janneshopman** [comment]
Hi,

Thank you for this suggestion. I would like to include this in my code. One problem in the function: turbulence->divDevRhoReff(U) is that it is calculated as (from lines 104-116 of https://www.openfoam.com/documentation/guides/latest/api/linearViscousStress_8C_source.html):

template<class BasicTurbulenceModel>
[Foam::tmp<Foam::fvVectorMatrix>](https://www.openfoam.com/documentation/guides/latest/api/classFoam_1_1tmp.html)
[Foam::linearViscousStress<BasicTurbulenceModel>::divDevRhoReff](https://www.openfoam.com/documentation/guides/latest/api/classFoam_1_1functionObjects_1_1momentumError.html#a9d2c6d79f4111210f4e399f55369a0f0)
(
[volVectorField](https://www.openfoam.com/documentation/guides/latest/api/classFoam_1_1GeometricField.html)& [U](https://www.openfoam.com/documentation/guides/latest/api/combustion_2PDRFoam_2pEqn_8H.html#a81cf6107131a3583e2b0b762cb9c2862)
) const
{
return
(
- [fvc::div](https://www.openfoam.com/documentation/guides/latest/api/namespaceFoam_1_1fvc.html#a585533948ad80abfe589d920c641f251)((this->alpha_*this->rho_*this->nuEff())*[dev2](https://www.openfoam.com/documentation/guides/latest/api/namespaceFoam.html#a01b360e0d7ad7b43a8037fc09929e96f)([T](https://www.openfoam.com/documentation/guides/latest/api/combustion_2coldEngineFoam_2createFieldRefs_8H.html#abb80dfe0d55ec4a3a97b2dc1353d66dc)([fvc::grad](https://www.openfoam.com/documentation/guides/latest/api/namespaceFoam_1_1fvc.html#a094a9cf8cbcd7511778abc82036c8a4b)([U](https://www.openfoam.com/documentation/guides/latest/api/combustion_2PDRFoam_2pEqn_8H.html#a81cf6107131a3583e2b0b762cb9c2862)))))
- [fvm::laplacian](https://www.openfoam.com/documentation/guides/latest/api/namespaceFoam_1_1fvm.html#a05679b154654e47e9f1e2fbd2bc0bae0)(this->alpha_*this->rho_*this->nuEff(), [U](https://www.openfoam.com/documentation/guides/latest/api/combustion_2PDRFoam_2pEqn_8H.html#a81cf6107131a3583e2b0b762cb9c2862))
);
},

in which spatial discretisations of fvc::div, fvc::grad, fvm::laplacian are taken via fvSchemes. It would be nice if it is directly hard-coded using the symmetry preserving div/grad/lap operators.

Rewriting the Laplacian in this example would result in:

UEqnSourceB = -
(
    - [fvc::div](https://www.openfoam.com/documentation/guides/latest/api/namespaceFoam_1_1fvc.html#a585533948ad80abfe589d920c641f251)((this->alpha_*this->rho_*this->nuEff())*[dev2](https://www.openfoam.com/documentation/guides/latest/api/namespaceFoam.html#a01b360e0d7ad7b43a8037fc09929e96f)([T](https://www.openfoam.com/documentation/guides/latest/api/combustion_2coldEngineFoam_2createFieldRefs_8H.html#abb80dfe0d55ec4a3a97b2dc1353d66dc)([fvc::grad](https://www.openfoam.com/documentation/guides/latest/api/namespaceFoam_1_1fvc.html#a094a9cf8cbcd7511778abc82036c8a4b)([U](https://www.openfoam.com/documentation/guides/latest/api/combustion_2PDRFoam_2pEqn_8H.html#a81cf6107131a3583e2b0b762cb9c2862)))))
    - fv::gaussLaplacianScheme<vector, scalar>
      (
          mesh,
          reverseLinear<scalar>(mesh),        // Scheme is not used
          fv::uncorrectedSnGrad<vector>(mesh)
      ).fvmLaplacian(turbulence->alpha_*turbulence->rho_*turbulence->nuEff(), U)
).


Although one issue would be that only the divDevReff function of the linearViscousStress model is now hard-coded. It might have a different function for different models, I do not know this.

By leaving the code as:

UEqnSourceB = - turbulence->divDevRhoReff(U);,


as you suggested, the code loses the hard-coded symmetry-preserving laplacian if no turbulence model is used.

One possible solution might be to use hard-coded laplacian in case no turbulence model is used, and leave the spatial discretisation to the user in case a model is used. This could be done by adding an if-statement or something similar, where my original code is used in absence of a model and your suggestion is used in case a model is selected.

In short:

I do agree that the solver is missing the turbulence model in this term, but I am not sure how to retain the hard-coded spatial discretisation if I use a solution similar to yours. I would be happy to discuss or hear your suggestions.

Kind regards,

Jannes

**Comment by janneshopman** [comment]
I added your suggestion including a switch, thanks

**Comment by maxreiss123** [comment]
Hi, many thanks.
